src/s_inter.c

```diff
@@ -138,6 +203,514 @@ void sys_set_temppath(void);
 void sys_set_extrapath(void);
 void sys_set_startup(void);
 void sys_stopgui(void);
+void sys_lockio();
+void sys_unlockio();
+static void fdp_select(fd_set *readset, struct timeval timout, const t_fdp_manager manager);
+
+#ifdef THREADED_IO
+#define RB_SIZE (8*INBUFSIZE)
+static void rb_dosend(ring_buffer*, int ignoreSigFd);
+static int rbsend_init();
+struct _rbskt { // t_rbskt
+    ring_buffer* rs_rb;
+    int rs_preserve_boundaries;
+    int rs_errno; // used by streaming sockets to report errors
+    int rs_closed; // used by streaming sockets to report the other end has gracefully closed the connection
+};
+
+static t_rbskt* rbskt_new(int preserve_boundaries) {
+    t_rbskt* x = (t_rbskt*)getbytes(sizeof(*x));
+    x->rs_rb = rb_create(RB_SIZE);
+    x->rs_preserve_boundaries = preserve_boundaries;
+    x->rs_errno = 0;
+    x->rs_closed = 0;
+    return x;
+}
+
+static void rbskt_free(t_rbskt* x) {
+    if(x) {
+        rb_free(x->rs_rb);
+        freebytes(x, sizeof(*x));
+    }
+}
+
+static int rbskt_ready(t_rbskt* x) {
+    return x && x->rs_rb && (x->rs_closed || x->rs_errno || rb_available_to_read(x->rs_rb));
+}
+
+int rbskt_bytes_available(t_rbskt* rbskt) {
+    return rb_available_to_read(rbskt->rs_rb);
+}
+
+// to ensure the correct order is preserved when calling send()/sendto()
+// directly from outside rb_dosendone(), always flush the corresponding
+// ringbuffer first
+// should_lock should be set to true if this is not called from the IO thread
+static void sendrb_flush(ring_buffer* rb, int should_lock) {
+    if(should_lock)
+        sys_lockio();
+    rb_dosend(rb, -1);
+    if(should_lock)
+        sys_unlockio();
+}
+
+// A sendto() that writes to a ring_buffer and is real-time safe
+static ssize_t rb_sendto(ring_buffer* rb, int socket, const void *buffer, size_t length, int flags, void* addr, size_t addrlen)
+{
+    //printf("rb_sendto: fd: %d, rb: %p \"%s\"\n", socket, rb, (char*)buffer);
+    size_t maxLength = rb_available_to_write(rb);
+    size_t desiredLength = sizeof(size_t) + length;
+    size_t actualLength = desiredLength <= maxLength ? desiredLength : maxLength;
+    if(actualLength < desiredLength)
+    {
+        error("rb buffer is full, sending packet from the audio thread");
+        sendrb_flush(rb, 1);
```

Comment:
If the ringbuffer ever gets full, it means it is too small. Maybe we could double the size after the flush?

Reply:
What I was thinking here is that if we are aiming for "best-effort real-time", then it's OK if there is an occasional very big packet going out at some point. Resizing could be fine, but should we set an upper limit? I should probably factor out the sizes of everything at the top of the file.

Reply:
> then it's OK if there is an occasional very big packet going out at some point.

The ringbuffer can also fill up if you send many small packets at once. I think a few mallocs are better than repeatingly locking and flushing. If the ringbuffer doubles in size, the appropriate size will be reached very quickly. I don't think there has to be an upper limit.

Reply:
b7c19988

A bit inelegant to pass around a `ring_buffer**`, but it's internal so it's kind of OK.

Reply:
I think it would be much more elegant if `ring_buffer` simply had a resize method ;-)

Reply:
Also, just instead of
```
// + 1 because one byte is always reserved
size_t size = rb_available_to_write(*rb) + 1;
```
just use `rb->size` :-)

Reply:
> just use rb->size :-)

Didn't realise it was public!

> I think it would be much more elegant if ring_buffer simply had a resize method ;-)

fair point. Probably the best way to go.

Reply:
>  think it would be much more elegant if ring_buffer simply had a resize method ;-)

here it is fdfd681f